### PR TITLE
Fix incorrect syntax for dropping primary indexes in PostgreSQL

### PR DESCRIPTION
**Pull request by f-lombardo**
|      Q       |   A
|------------- | -----------
| Type         | bug/feature/improvement
| Fixed issues | https://github.com/doctrine/dbal/issues/6024

#### Summary

The correct way of dropping primary index in PostgreSQL is not dropping the index, but removing the primary key constraints

### CODE REVIEW

File: tests/Platforms/PostgreSQLPlatformTest.php [hunk 1]
```diff
@@ -581,6 +581,27 @@ public function testDroppingConstraintsBeforeColumns(): void
...
+    public function testDroppingPrimaryKey(): void
+    {
+        $oldTable = new Table('mytable');
+        $oldTable->addColumn('id', 'integer');
+        $oldTable->setPrimaryKey(['id']);
+
+        $newTable = clone $oldTable;
+        $newTable->dropPrimaryKey();
+
+        $diff = (new Comparator())->diffTable($oldTable, $newTable);
```
Comment (phansys) [inline tests/Platforms/PostgreSQLPlatformTest.php]: The `diffTable()` method is deprecated. You should use `compareTables()` instead.

File: src/Platforms/PostgreSQLPlatform.php [hunk 2]
```diff
@@ -809,6 +809,18 @@ public function getDropForeignKeySQL($foreignKey, $table)
         return $this->getDropConstraintSQL($foreignKey, $table);
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    public function getDropIndexSQL($index, $table = null)
+    {
+        if ($index->isPrimary()) {
+            return $this->getDropConstraintSQL($table.'_pkey', $table);
```
Comment (phansys) [inline src/Platforms/PostgreSQLPlatform.php]: What if the constraint is not using the default name? https://www.postgresql.org/docs/current/sql-createtable.html#id-1.9.3.85.1.

**Review by phansys — CHANGES_REQUESTED** [review]

**Comment by derrabus** [comment]
Please add a functional test that reproduces your problem. Your test makes assertions on the produced SQL statements, but that neither tells me that the new statement is correct nor that the old one is incorrect.

**Comment by f-lombardo** [comment]
> Please add a functional test that reproduces your problem. Your test makes assertions on the produced SQL statements, but that neither tells me that the new statement is correct nor that the old one is incorrect.

Yes, it would be a good thing to add a functional test. Could you please give me a pointer to the process of creating a new functional test?

**Comment by derrabus** [comment]
Have a look at any test in the `tests/Functional` directory.

**Comment by f-lombardo** [comment]
> Have a look at any test in the `tests/Functional` directory.

I've already seen them and they seems simple to replicate. My problem is that I don't know exactly how to run them against PostgreSQL.

Thanks in advance.

**Comment by derrabus** [comment]
You'll find a PHPUnit configuration in `ci/github/phpunit/pgsql.xml`. Prepare a local Postgres server with an empty database and adjust the settings in `ci/github/phpunit/pgsql.xml` to match your environment.

File: src/Platforms/PostgreSQLPlatform.php [hunk 3]
```diff
@@ -809,6 +809,22 @@ public function getDropForeignKeySQL($foreignKey, $table)
...
 
+    /**
+     * {@inheritDoc}
+     */
+    public function getDropIndexSQL($index, $table = null)
+    {
+        if ($index instanceof Index && $index->isPrimary()) {
+            $constraintName = 'primary' === $index->getName() ? $table . '_pkey' : $index->getName();
+            return $this->getDropConstraintSQL($constraintName, $table);
+        } elseif (is_string($index) && '"primary"' === $index) {
```
Comment (phansys) [inline src/Platforms/PostgreSQLPlatform.php]: There is no need for an `else` condition when the previous one breaks the execution flow with a `return` statement.
Additionally, I think there is no need for the `is_string()` check here.

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        $table = new Table('mytable');
+        $table->addColumn('id', 'integer');
+        $table->setPrimaryKey(['id']);
+        $this->dropAndCreateTable($table);
+
+        $newTable = clone $table;
+        $newTable->dropPrimaryKey();
+
+        xdebug_break();
```
Comment (phansys) [inline tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: :eyes:

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        $newTable->dropPrimaryKey();
+
+        $diff = (new Comparator())->compareTables($table, $newTable);
+
+        $statements = $this->connection->getDatabasePlatform()->getAlterTableSQL($diff);
+        foreach ($statements as $statement) {
+            $this->connection->executeStatement($statement);
+        }
+
+        $this->assertTrue($this->count($statements) > 0);
```
Comment (derrabus) [inline tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: This assertion only tests that _something_ has been executed on the database. We should introspect the table and make sure that it does not have a primary key anymore.

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+/** @requires extension pdo_pgsql */
+class DBAL6024Test extends FunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        if (TestUtil::isDriverOneOf('pdo_pgsql')) {
+            return;
+        }
+
+        self::markTestSkipped('This test requires the pdo_pgsql driver.');
```
Comment (derrabus) [inline tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: Is there a reason why we should not run this test on all platforms? It looks pretty generic to me.

**Review by derrabus — CHANGES_REQUESTED** [review]
Thanks. We're almost there, I think. 🙂

**Comment by derrabus** [comment]
You can run `vendor/bin/phpcbf` to fix the issues reported by PHPCS. And please have a look at the errors reported by PHPStan. They're valid as far as I can tell.

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Tests\FunctionalTestCase;
+
+/**
+ * How to run this test with PostgreSQL (the original target of the issue):
+ * 1) Start a PostgreSQLInstance
+ * 2) If needed, change ci/github/phpunit/pdo_pgsql.xml according to your PostgreSQL local settings
+ * 3) Run
+ * vendor/bin/phpunit -c ci/github/phpunit/pdo_pgsql.xml tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php
+ */
```
Comment (derrabus) [inline tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: This applies to all tests. Generic instructions on how to run the test suite could be added to the documentation, but they don't belong here.

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+/** @requires extension pdo_pgsql */
```
Comment (derrabus) [inline tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: Should be removed.

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php (hunk 7 above)
Comment (f-lombardo) [reply tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: I agree. Where is the best place to add this kind of instruction?

Comment (derrabus) [reply tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: Somewhere in the documentation, I suppose. There's a "docs" folder in this repository. But that should be a separate PR, please.

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        $newTable->dropPrimaryKey();
+
+        $diff = (new Comparator())->compareTables($table, $newTable);
+
+        $statements = $this->connection->getDatabasePlatform()->getAlterTableSQL($diff);
+        foreach ($statements as $statement) {
+            $this->connection->executeStatement($statement);
+        }
+
+        $schemaManager = $this->connection->createSchemaManager();
```
Comment (derrabus) [inline tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: Sorry, I missed that you're calling `new Comparator()` directly. Let's use `createComparator()` instead.

```suggestion
        $schemaManager = $this->connection->createSchemaManager();
        $diff          = $schemaManager->createComparator()->compareTables($table, $newTable);

        $statements = $this->connection->getDatabasePlatform()->getAlterTableSQL($diff);
        foreach ($statements as $statement) {
            $this->connection->executeStatement($statement);
        }
```

**Review by derrabus — CHANGES_REQUESTED** [review]
Your new test apparently only works on MySQL and Postgres. That's a problem that we can solve in a follow-up PR. Please skip the test for all failing platforms. Sorry for the back and forth here. 🙈

**Comment by f-lombardo** [comment]
> Your new test apparently only works on MySQL and Postgres. That's a problem that we can solve in a follow-up PR. Please skip the test for all failing platforms. Sorry for the back and forth here. 🙈

Uhm, I think it would be better to revert to the initial code that was designed just for PostgreSQL, since the fix is actually realated to a problem with this DB. What do you think?

File: tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Tests\FunctionalTestCase;
+use Doctrine\DBAL\Tests\TestUtil;
+
+/** @requires extension pdo_pgsql */
+class DBAL6024Test extends FunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        if (TestUtil::isDriverOneOf('pdo_pgsql')) {
```
Comment (derrabus) [inline tests/Functional/Driver/PDO/PgSQL/DBAL6024Test.php]: The test should be run with the `pgsql` driver as well.

File: src/Platforms/PostgreSQLPlatform.php [hunk 11]
```diff
@@ -809,6 +810,29 @@ public function getDropForeignKeySQL($foreignKey, $table)
         return $this->getDropConstraintSQL($foreignKey, $table);
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    public function getDropIndexSQL($index, $table = null)
+    {
+        /** @psalm-suppress MissingClosureParamType */
+        $tableName = static fn ($t): string => $t instanceof Table ? $t->getName() : (string) $t;
```
Comment (derrabus) [inline src/Platforms/PostgreSQLPlatform.php]: Let's make this a private method. This allows us to document the parameter type via PHPDoc which should make Psalm happy.

**Review by derrabus — APPROVED** [review]

**Pull request merged**